### PR TITLE
Lints

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Adds various lints and fixes our implementations. This ensures items are always documented and we don't miss any in CI. Thanks!

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
I've disabled the router bench temporarily; we don't really use it much anymore, but can always spin it back up if we want to.

**Comment by yoshuawuyts** [comment]
And I've restored the router bench. All `Server::respond` does is clone various things and then call the router. This should be close enough to the original bench.

**Review by Fishrock123 — APPROVED** [review]

File: src/lib.rs [hunk 1]
```diff
@@ -194,10 +194,9 @@
 //! for Async Rust. We have a long journey ahead of us. But we're excited you're here with us!
 
 #![cfg_attr(feature = "docs", feature(doc_cfg))]
-// #![warn(missing_docs)]
-#![warn(missing_debug_implementations, rust_2018_idioms)]
-#![doc(test(attr(deny(rust_2018_idioms, warnings))))]
-#![doc(test(attr(allow(unused_extern_crates, unused_variables))))]
+#![forbid(unsafe_code, rust_2018_idioms)]
+#![deny(missing_debug_implementations, nonstandard_style)]
+#![warn(missing_docs, unreachable_pub, future_incompatible)]
```
Comment (taiki-e) [inline src/lib.rs:199]:
```suggestion
#![forbid(unsafe_code)]
#![deny(missing_debug_implementations, nonstandard_style)]
#![warn(missing_docs, unreachable_pub, future_incompatible, rust_2018_idioms)]
```

As said in https://github.com/taiki-e/pin-project/issues/281#issuecomment-699543730, I do not recommend using `forbid` on the lints other than `unsafe_code` lint.

File: src/lib.rs [hunk 2]
```diff
@@ -194,10 +194,9 @@
 //! for Async Rust. We have a long journey ahead of us. But we're excited you're here with us!
 
 #![cfg_attr(feature = "docs", feature(doc_cfg))]
-// #![warn(missing_docs)]
-#![warn(missing_debug_implementations, rust_2018_idioms)]
-#![doc(test(attr(deny(rust_2018_idioms, warnings))))]
-#![doc(test(attr(allow(unused_extern_crates, unused_variables))))]
```
Comment (taiki-e) [inline src/lib.rs:200]: These two lines are needed to [catch the doc test warnings](https://doc.rust-lang.org/nightly/rustdoc/the-doc-attribute.html#testattr). (Normal crate-level lints cannot catch doc test warnings.) So you probably don't need to remove them.

**Comment by Fishrock123** [comment]
See also https://github.com/http-rs/async-sse/pull/11

**Pull request merged**